### PR TITLE
ENH: Make datasets have subclass relationships

### PR DESCRIPTION
**Pull request by llllllllll**
Needed for the EarningsCalendar work we are doing

### CODE REVIEW

File: zipline/pipeline/data/dataset.py [hunk 1]
```diff
@@ -119,5 +140,5 @@ def __repr__(self):
         return '<DataSet: %r>' % self.__name__
 
 
-class DataSet(with_metaclass(DataSetMeta)):
+class DataSet(with_metaclass(DataSetMeta, object)):
```
Comment (ssanderson) [inline zipline/pipeline/data/dataset.py]: Was this previously an old-style class in PY2?

File: zipline/pipeline/data/dataset.py [hunk 2]
```diff
@@ -97,20 +114,24 @@ class DataSetMeta(type):
...
+                column_names.add(maybe_colname)
 
-        newtype._columns = frozenset(_columns)
+        newtype._column_names = frozenset(column_names)
         return newtype
 
     @property
     def columns(self):
-        return self._columns
+        return frozenset(
```
Comment (ssanderson) [inline zipline/pipeline/data/dataset.py]: Should this be a lazyval now that it's doing semi-nontrivial work?

File: zipline/pipeline/data/dataset.py (hunk 1 above)
Comment (llllllllll) [inline zipline/pipeline/data/dataset.py]: yeah, I believe calling `type('name', (), {})` creates an old style class. At least it excludes `object` from the `__mro__`.

File: zipline/pipeline/data/dataset.py (hunk 2 above)
Comment (llllllllll) [inline zipline/pipeline/data/dataset.py]: The cost to compute this is still trivial. Idk if we should jump to caching just yet.

File: zipline/pipeline/data/dataset.py [hunk 3]
```diff
@@ -97,20 +114,24 @@ class DataSetMeta(type):
     """
 
     def __new__(mcls, name, bases, dict_):
-        newtype = type.__new__(mcls, name, bases, dict_)
-        _columns = []
+        newtype = super(DataSetMeta, mcls).__new__(mcls, name, bases, dict_)
+        column_names = set().union(
```
Comment (ssanderson) [inline zipline/pipeline/data/dataset.py]: Might be worth adding a comment here like `# Get columns from any parent classes`.

File: zipline/pipeline/data/dataset.py [hunk 4]
```diff
@@ -97,20 +114,24 @@ class DataSetMeta(type):
...
+            *(getattr(base, '_column_names', ()) for base in bases)
+        )
         for maybe_colname, maybe_column in iteritems(dict_):
             if isinstance(maybe_column, Column):
-                bound_column = maybe_column.bind(newtype, maybe_colname)
-                setattr(newtype, maybe_colname, bound_column)
-                _columns.append(bound_column)
+                bound_column_descr = maybe_column.bind(maybe_colname)
```
Comment (ssanderson) [inline zipline/pipeline/data/dataset.py]: Might be worth commenting here something like `# Add columns from the newly-created class.`.

File: tests/pipeline/test_term.py [hunk 5]
```diff
@@ -321,3 +325,31 @@ class SomeFactorNoDType(SomeFactor):
...
+
+class SubDataSetTestCase(TestCase):
+    def test_subdataset(self):
+        some_dataset_map = {
+            column.name: column for column in SomeDataSet.columns
+        }
+        sub_dataset_map = {
+            column.name: column for column in SubDataSet.columns
+        }
+        self.assertEqual(
```
Comment (ssanderson) [inline tests/pipeline/test_term.py]: I was confused by this assertion when reading.  I think it'd be clearer to assert that `SomeDataSet.columns` is equal to `SubDataSet.columns`, which I think it both equivalent and closer to the intent of what's being tested here.

File: tests/pipeline/test_term.py [hunk 6]
```diff
@@ -28,6 +28,10 @@ class SomeDataSet(DataSet):
     buzz = Column(float64_dtype)
 
 
+class SubDataSet(SomeDataSet):
```
Comment (ssanderson) [inline tests/pipeline/test_term.py]: Can we add a test case for adding a new column in a subdataset?

File: zipline/pipeline/data/dataset.py (hunk 1 above)
Comment (ssanderson) [inline zipline/pipeline/data/dataset.py]: :scream_cat:

**Comment by ssanderson** [comment]
By and large LGTM.  Had a couple requests for comments and one request for a testcase.

File: tests/pipeline/test_term.py (hunk 5 above)
Comment (llllllllll) [inline tests/pipeline/test_term.py]: Sounds good, this is what I am trying to test.

**Comment by llllllllll** [comment]
added the test and comments. merging on pass

**Pull request merged**